File: HW02_B0829002/partition2/app_transaction.go
package main

import (
	"crypto/sha256"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
)

// checkBlock determines the current block by checking the existence of block files.
func checkBlock() int {
	i := 1
	for {
		if _, err := os.Stat(strconv.Itoa(i) + ".txt"); os.IsNotExist(err) {
			break
		}
		i++
	}
	return i
}

// writeBlock writes the transaction to the specified block.
func writeBlock(currentBlock int, sender, receiver, amount string) {
	fileName := fmt.Sprintf("%d.txt", currentBlock)
	f, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer f.Close()
	_, err = f.WriteString(fmt.Sprintf("%s, %s, %s\n", sender, receiver, amount))
	if err != nil {
		fmt.Println("Error writing to file:", err)
		return
	}
	fmt.Printf("Transaction written to block %s\n", fileName)
}

func main() {
	if len(os.Args) != 4 {
		fmt.Println("Usage: go run app_transaction.go sender receiver amount")
		os.Exit(1)
	}

	sender := os.Args[1]
	receiver := os.Args[2]
	amount := os.Args[3]

	currentBlock := checkBlock()
	fmt.Printf("Current block: %d.txt\n", currentBlock-1)

	fileName := fmt.Sprintf("%d.txt", currentBlock-1)
	data, err := ioutil.ReadFile(fileName)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	// if file length is greater than 8, create a new block
	lines := 0
	for _, b := range data {
		if b == '\n' {
			lines++
		}
	}
	

	if lines >= 7 {
		text := string(data)
		fmt.Println("Text: ", text)
		hash := sha256.Sum256([]byte(text))
		hashStr := fmt.Sprintf("%x", hash)
		fmt.Println("Hash: ", hashStr)

		newFileName := fmt.Sprintf("%d.txt", currentBlock)
		f, err := os.Create(newFileName)
		if err != nil {
			fmt.Println("Error creating file:", err)
			return
		}
		defer f.Close()
		f.WriteString(fmt.Sprintf("Sha256 of previous block: %s\n", hashStr))
		f.WriteString(fmt.Sprintf("Next block: %d.txt\n", currentBlock+1))
		writeBlock(currentBlock, sender, receiver, amount)
	} else {
		writeBlock(currentBlock-1, sender, receiver, amount)
	}
}

File: HW01/app_transaction.py

import hashlib
from library import checkBlock
# arg 1: sender, arg 2: receiver, arg 3: amount

# get arg from python3 app_transaction.py sender receiver amount
import sys
try:
    sender = sys.argv[1]
    receiver = sys.argv[2]
    amount = sys.argv[3]
except:
    print("Usage: python3 app_transaction.py sender receiver amount")
    sys.exit()
if len(sys.argv) != 4:
    print("Usage: python3 app_transaction.py sender receiver amount")
    sys.exit()

        
def writeBlock(currentBlock, sender, receiver, amount):
    f = open("{}.txt".format(currentBlock), "a")
    f.write("{}, {}, {}\n".format(sender, receiver, amount))
    f.close()
    print("Transaction written to block {}.txt".format(currentBlock))
        
currentBlock = checkBlock()
print("Current block: {}.txt".format(currentBlock-1))
# check the current block >= 6 items, if so, create a new block
f = open("{}.txt".format(currentBlock-1), "r")
lines = f.readlines()
if len(lines) >= 8:
    # get all text from the current block and hash it
    f = open("{}.txt".format(currentBlock-1), "r")
    lines = f.readlines()
    # print lines
    text = ""
    for line in lines:
        text += line
    # hash the text
    print("Text: ", text)
    hash_object = hashlib.sha256(text.encode())
    print("Hash: ", hash_object.hexdigest())
    # open a new block
    f = open("{}.txt".format(currentBlock), "w+")
    f.write("Sha256 of previous block: {}\n".format(hash_object.hexdigest()))
    f.write("Next block: {}.txt\n".format(currentBlock+1))
    writeBlock(currentBlock, sender, receiver, amount)
else:
    writeBlock(currentBlock-1, sender, receiver, amount)
    
    